### PR TITLE
Add compressedLogSize, uncompressedLogSize and logEventIdx metadata to every file synchronization request.

### PR DESCRIPTION
**Pull request by jackluo923**
# Description
During processing of asynchronous file synchronization requests, custom metadata may be gathered at the time of submitting the synchronization request to make decisions. This commit enables the functionality to allow downstream appenders to customize the metadata that is included with every synchronization request.

# Validation performed
All existing unit tests passed.

### CODE REVIEW

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java [hunk 1]
```diff
@@ -504,11 +513,13 @@ private class SyncRequest extends Request {
       public final String logFileBaseName;
       public final long logRolloverTimestamp;
       public final boolean deleteFile;
+      public final Map<String, Object> metadata;
 
       public SyncRequest (String logFileBaseName, long logRolloverTimestamp, boolean deleteFile) {
         this.logFileBaseName = logFileBaseName;
         this.logRolloverTimestamp = logRolloverTimestamp;
         this.deleteFile = deleteFile;
+        this.metadata = computeSyncRequestMetadata();
```
Comment (kirkrodrigues) [inline src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: This seems a bit messy:
* We're depending on SyncRequest being constructed from the parent class here whereas that might not always be the case in which case the request metadata may be for a different file than the one that the sync request is for.
* It breaks the convention we've already established where all the members are passed in.

I would suggest simply passing in the metadata as a parameter and adding a call to computeSyncRequestMetadata at every callsite of addSyncRequest.

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java [hunk 2]
```diff
@@ -356,6 +357,14 @@ protected abstract void sync (String baseName, long logRolloverTimestamp, boolea
    */
   protected abstract String computeLogFileName (String baseName, long logRolloverTimestamp);
 
+  /**
+   * Computes the metadata which will be included in a synchronization request
+   * @return The computed metadata map
+   */
+  protected Map<String, Object> computeSyncRequestMetadata () {
+    return new HashMap<>();
```
Comment (kirkrodrigues) [inline src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: Couldn't we return `null`?

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java [hunk 3]
```diff
@@ -504,11 +513,13 @@ private class SyncRequest extends Request {
       public final String logFileBaseName;
       public final long logRolloverTimestamp;
       public final boolean deleteFile;
+      public final Map<String, Object> metadata;
```
Comment (kirkrodrigues) [inline src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: You're not using this anywhere?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java (hunk 2 above)
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: Sure.

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java (hunk 3 above)
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: There are currently no concrete classes in this library that uses the metadata. However, downstream concrete appenders extending the abstract class will use them.

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java [hunk 4]
```diff
@@ -490,9 +503,11 @@ public void addShutdownRequest () {
      * @param logRolloverTimestamp The approximate timestamp of when the target
      * log file was rolled over
      * @param deleteFile Whether the log file can be deleted after syncing.
+     * @param metadata The metadata map object to be included in the sync request.
```
Comment (kirkrodrigues) [inline src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: `Metadata to synchronize with the file`?

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java (hunk 3 above)
Comment (kirkrodrigues) [reply src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: How are they going to access them?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java (hunk 4 above)
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: Modified to `Extra metadata for the file`

File: src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java (hunk 3 above)
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java]: Good point. Made changes to `sync` method to allow downstream appenders to use it.

**Review by kirkrodrigues — APPROVED** [review]

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Where is `logEventIdx` being incremented?

**Comment by jackluo923** [comment]
> Where is `logEventIdx` being incremented?

In the latest commit, `logEventIdx` is now moved down to `AbstractBufferedRollingFileAppender` and incremented in the `append` method.

**Review by kirkrodrigues — APPROVED** [review]

**Pull request merged**